cpp/src/join/semi_join.cu

```diff
@@ -36,26 +36,174 @@
+  for (auto const& col : table) {
+    accumulate_nested_columns(
+      table_view{std::vector<column_view>{col.child_begin(), col.child_end()}}, result);
+  }
+}
+
 }  // namespace
 
+rmm::device_uvector<bool> left_semi_join_contains(table_view const& left_keys,
+                                                  table_view const& right_keys,
+                                                  null_equality compare_nulls,
+                                                  nan_equality compare_nans,
+                                                  rmm::cuda_stream_view stream,
+                                                  rmm::mr::device_memory_resource* mr)
+{
+  // Use a hash map with key type is row hash values and map value type is `rhs_index_type` to store
+  // all indices of row in the rhs table.
+  using hash_map_type =
+    cuco::static_multimap<hash_value_type,
+                          rhs_index_type,
+                          cuda::thread_scope_device,
+                          rmm::mr::stream_allocator_adaptor<default_allocator<char>>,
+                          cuco::double_hashing<DEFAULT_JOIN_CG_SIZE, hash_type, hash_type>>;
+
+  auto map = hash_map_type(compute_hash_table_size(right_keys.num_rows()),
+                           cuco::sentinel::empty_key{std::numeric_limits<hash_value_type>::max()},
+                           cuco::sentinel::empty_value{rhs_index_type{cudf::detail::JoinNoneValue}},
+                           stream.value(),
+                           detail::hash_table_allocator_type{default_allocator<char>{}, stream});
+
+  auto const lhs_has_nulls = has_nested_nulls(left_keys);
+  auto const rhs_has_nulls = has_nested_nulls(right_keys);
+
+  // Insert all row hash values and indices of the rhs table.
+  {
+    auto const hasher   = cudf::experimental::row::hash::row_hasher(right_keys, stream);
+    auto const d_hasher = hasher.device_hasher(nullate::DYNAMIC{rhs_has_nulls});
+
+    auto const kv_it = cudf::detail::make_counting_transform_iterator(
+      size_type{0},
+      make_pair_fn<rhs_index_type, decltype(d_hasher)>{d_hasher, map.get_empty_key_sentinel()});
+
+    // If right table has nulls but they are compared unequal, don't insert them.
+    // Otherwise, it was known to cause performance issue:
+    // - https://github.com/rapidsai/cudf/pull/6943
+    // - https://github.com/rapidsai/cudf/pull/8277
+    if (rhs_has_nulls && compare_nulls == null_equality::UNEQUAL) {
+      // Gather all columns at all levels from the right table.
+      // The columns without having nulls will not be considered in `bitmask_and`.
+      auto const right_columns_with_nulls = [&] {
+        auto result = std::vector<column_view>{};
+        accumulate_nested_columns(right_keys, result);
+        return result;
+      }();
+
+      [[maybe_unused]] auto const [row_bitmask, tmp] =
+        cudf::detail::bitmask_and(table_view{right_columns_with_nulls}, stream);
```

Comment:
I'm not sure this logic is correct. I recall there being a difference in top-level vs nested nulls. @PointKernel ran into this with the groupby refactor recently. 

Yunsong, what did we decide there?

Reply:
`bitmask_and` denotes top-level nulls only and doesn't check nested nulls. So I'm not sure it's still a valid predicate for `insert_if` or any other `*_if` functions.

Reply:
Here all the (nullable) columns at all nested levels are gathered to compute `bitmask_and` so the logic is still correct.

Reply:
IIRC, top-level nulls and nested nulls are not the same "null" in spark (e.g. https://github.com/rapidsai/cudf/pull/10770#discussion_r878252863). We may want to check the expected behavior in spark and try to be consistent.

Reply:
I'm not sure how to handle this. But the old implementation just flattens all the nested columns (using struct utility function) and computes `bitmask_and` thus I'm just doing exactly the same here.